rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoJournaller.scala

```diff
@@ -32,73 +38,97 @@ class RxMongoJournaller(driver: RxMongoDriver) extends MongoPersistenceJournalli
       val collected = batch.toStream.collect { case Success(doc) => doc }
-      val result = Failover2(driver.connection, driver.failoverStrategy) { () =>
-        for {
-          j <- journal.bulkInsert(collected, ordered = true, writeConcern).map(_ => batch.map(_.map(_ => ())))
-          _  <- if(driver.realtimeEnablePersistence)
-                  realtime.bulkInsert(collected, ordered = true, writeConcern).map(_ => zero).recover {
-                    case t: Throwable =>
-                      logger.error("Error bulk inserting into realtime collection", t)
-                      zero
-                  }
-                else Future.successful(zero)
-        } yield j
+      Failover2(driver.connection, driver.failoverStrategy) { () =>
+        journal.bulkInsert(collected, ordered = true, writeConcern).map(_ => batch.map(_.map(_ => ())))
       }.future
+    } else {
+      Future.sequence(batch.map {
+        case Success(document: BSONDocument) =>
+          journal.insert(document, writeConcern).map(writeResultToUnit)
+        case f: Failure[_] => Future.successful(Failure[Unit](f.exception))
+      })
+    }
+  }
+
+  private[this] def doBatchRealtimeAppend(writes: ISeq[AtomicWrite])(implicit ec: ExecutionContext): Future[ISeq[Try[Unit]]] = {
+    val batch = writes.map(aw => Try(driver.serializeJournal(Atom[BSONDocument](aw, driver.useLegacySerialization))))
+    val zero = ISeq.empty[Try[Unit]]
 
-      result
+    if (batch.forall(_.isSuccess)) {
+      val collected = batch.toStream.collect { case Success(doc) => doc }
+      Failover2(driver.connection, driver.failoverStrategy) { () =>
+          realtime.bulkInsert(collected, ordered = true, writeConcern).map(_ => zero).recover {
+              case t: Throwable =>
+                logger.error("Error bulk inserting into realtime collection", t)
+                zero
+        }.map(_ => zero)
+      }.future
     } else {
       Future.sequence(batch.map {
         case Success(document: BSONDocument) =>
-          for {
-            j <- journal.insert(document, writeConcern).map(writeResultToUnit)
-            _ <- if (driver.realtimeEnablePersistence)
-                    realtime.insert(document, writeConcern).map(_ => zero).recover {
-                      case t: Throwable =>
-                        logger.error("Error inserting into realtime collection", t)
-                        zero
-                    }
-                  else Future.successful(zero)
-          } yield j
+          realtime.insert(document, writeConcern).map(_ => zero).recover {
+            case t: Throwable =>
+              logger.error("Error inserting into realtime collection", t)
+              zero
+          }.map(_ => Try[Unit](()))
         case f: Failure[_] => Future.successful(Failure[Unit](f.exception))
       })
     }
+
+  }
+
+  private[mongodb] override def batchAppend(writes: ISeq[AtomicWrite])(implicit ec: ExecutionContext): Future[ISeq[Try[Unit]]] = {
```

Comment:
Done, but as codes were slightly different in `doBatchJournalAppend` and `doRealtimeAppend` methods, I kept two versions of the new `doBatchAppend` method for you to choose which one you prefer...